File: frontend/demo/component/select/select-complex-value-label.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/select';
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import type { SelectItem } from '@vaadin/select';
import { getPeople } from 'Frontend/demo/domain/DataService';
import { applyTheme } from 'Frontend/demo/theme';

@customElement('select-complex-value-label')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  @state()
  private items: SelectItem[] = [];

  protected override async firstUpdated() {
    const { people } = await getPeople({ count: 5 });
    // tag::snippet[]
    this.items = people.map((person) => ({
      label: `${person.firstName} ${person.lastName}`,
      value: `${person.id}`,
    }));
    // end::snippet[]
  }

  protected override render() {
    return html`<vaadin-select label="Assignee" .items="${this.items}"></vaadin-select>`;
  }
}

File: src/main/java/com/vaadin/demo/component/select/SelectComplexValueLabel.java
package com.vaadin.demo.component.select;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.DataService;
import com.vaadin.demo.domain.Person;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.select.Select;
import com.vaadin.flow.router.Route;

import java.util.List;

@Route("select-complex-value-label")
public class SelectComplexValueLabel extends Div {

  public SelectComplexValueLabel() {
    // tag::snippet[]
    Select<Person> select = new Select<>();
    select.setLabel("Assignee");
    // Display full name of the person as item text and selected value label
    select.setItemLabelGenerator(Person::getFullName);

    List<Person> people = DataService.getPeople(5);
    select.setItems(people);
    // end::snippet[]

    add(select);
  }

  public static class Exporter extends DemoExporter<SelectComplexValueLabel> { // hidden-source-line
  } // hidden-source-line
}
